futures-util/src/future/shared.rs

```diff
@@ -133,13 +134,28 @@ where
         debug_assert!(self.waker_key != NULL_WAKER_KEY);
     }
 
+    /// Safety: callers must first ensure that `self.inner.state`
+    /// is `COMPLETE`
+    unsafe fn take_or_clone_output(&mut self) -> Fut::Output {
+        if let Some(inner) = Arc::get_mut(&mut self.inner) {
```

Comment:
This is slightly racey because it will still clone the last element if other futures have completed but not yet been dropped. One option to prevent this would be to store `self.inner` in an `Option` and drop it on completion. WDYT?

Reply:
I think that is a good idea, the extra branch is unfortunate but it ensures that `Shared` will only clone the output once per completed instance of `Shared` (-1 for the original value).

I haven't documented the guarantee though.